loadbalancer: Add hostname support for ring hash and maglev

inf-rno opened the pull request:
Description:
- add support for using host->hostname() instead of resolvedAddress()
- this uses an additional field in common_lb_config to use_hostname
- throw an explicit EnvoyException if if this config is set on a cluster
  that is not StrictDNS or LogicalDNS
- fix the implementation of ringHashLB to use absl::InlinedVector without
  a size limit on the address

Risk Level: Low
Testing: Unit test
Docs Changes: Docs inline with API Protobuf
Release Notes: Yes
Fixes #9074

mattklein123 commented on docs/root/intro/version_history.rst:
```diff
@@ -17,6 +17,7 @@ Version history
 * http: fixing a bug in HTTP/1.0 responses where Connection: keep-alive was not appended for connections which were kept alive.
 * listener filters: listener filter extensions use the "envoy.filters.listener" name space. A
   mapping of extension names is available in the :ref:`deprecated <deprecated>` documentation.
+* loadbalancing: added support for using hostname based loadbalancing via `common_lb_config:use_hostname`.
```
Please ref link the actual config value.

mattklein123 commented on api/envoy/api/v2/cluster.proto:
```diff
@@ -438,6 +438,11 @@ message Cluster {
+    // If set to `true`, the cluster will use hostname instead of the resolved
+    // name to load balance hosts. Only valid for StrictDNS and LogicalDNS
+    // clusters.
+    bool use_hostname = 7;
```
Can we flesh out this documentation more and clarify what this actually does and that it actually has to do with hashing? I would pretty clearly indicate that it only works for hashing load balancers and provide some examples. Perhaps we should also call this `use_hostname_for_consistent_hashing` or something like that? WDYT?

mattklein123 commented on source/common/upstream/cluster_factory_impl.cc:
```diff
@@ -57,6 +57,13 @@ std::pair<ClusterSharedPtr, ThreadAwareLoadBalancerPtr> ClusterFactoryImplBase::
+
+  if (cluster.common_lb_config().use_hostname() &&
+      !(cluster.type() == envoy::config::cluster::v3::Cluster::STRICT_DNS ||
+        cluster.type() == envoy::config::cluster::v3::Cluster::LOGICAL_DNS)) {
```
Does this even make sense for logical_dns? I can't see anyone using logical DNS with a hashing LB as there is only ever a single logical host? Perhaps this should only be an option for strict DNS and we should actually have a specific strict DNS configuration of some type vs. having the config be global to all lb options? WDYT?

inf-rno replied:
My understanding might be a bit flawed but here's what I'm thinking:

This makes the most sense for StrictDNS which returns just a single IP address. It doesn't quite make sense for StrictDNS with multiple IP addresses because the Cluster will have multiple hosts but the same hostname. So the ring/maglev table will have multiple entries with the same hash but different IP. 

For LogicalDNS with single IP this works well as expected although should probably use StrictDNS. For LogicalDNS with multiple/changing IPs because there is only one host in envoy this can still work if the IP is rotated for failover reasons but not if it is rotated for loadbalancing. Still seems like it doesn't make sense here.

If we restrict it to just StrictDNS, where would you want the config to be? There isn't something like this afaik. I still think common_lb_config makes the most sense because it is common to the hashing loadbalancers and this config is passed into the constructors which simplifies the implementation.

inf-rno replied:
Sure, that makes sense. But if we are going to add another message then what about doing what I did in the previous attempt? Add a `MaglevLbConfig` to compliment the `RingHashLbConfig` and both have the `use_hostname_for_consistent_hashing` as options within them.

I'm good with either solution. Let me know what you prefer and I'll make it happen.

mattklein123 replied:
> I'm good with either solution. Let me know what you prefer and I'll make it happen.

I could go either way on this. In some sense a common config for all hashing load balancers is more future proof so that would be my mild preference?

mattklein123 replied:
One option would be to add the MaglevLbConfig, then also add a `CommonConsistentHashingLbConfig` and include that from both configs? Up to you.

mattklein123 replied:
Sorry where did we land here about supporting this for LogicalDNS? I'm still not convinced this makes sense? Should we only support this for StrictDNS where the user supplies  a set of host names that they want to route to?

mattklein123 commented on source/common/upstream/maglev_lb.h:
```diff
@@ -81,14 +82,15 @@ class MaglevLoadBalancer : public ThreadAwareLoadBalancerBase {
   Stats::ScopePtr scope_;
   MaglevLoadBalancerStats stats_;
   const uint64_t table_size_;
+  const bool use_hostname_;
```
Same comment in the code as the config. Can you make this variable more descriptive/longer based on what we decide the config name should be?

mattklein123 commented on source/common/upstream/maglev_lb.cc:
```diff
@@ -26,7 +26,7 @@ MaglevTable::MaglevTable(const NormalizedHostWeightVector& normalized_host_weigh
   for (const auto& host_weight : normalized_host_weights) {
     const auto& host = host_weight.first;
-    const std::string& address = host->address()->asString();
+    const std::string& address = use_hostname ? host->hostname() : host->address()->asString();
```
After this line can you ASSERT that address is not empty? Same in the ring hash case.

mattklein123 commented on source/common/upstream/ring_hash_lb.cc:
```diff
@@ -122,37 +124,31 @@ RingHashLoadBalancer::Ring::Ring(const NormalizedHostWeightVector& normalized_ho
-          offset_start +
-          StringUtil::itoa(hash_key_buffer + offset_start, StringUtil::MIN_ITOA_OUT_LEN, i);
-      absl::string_view hash_key(hash_key_buffer, total_hash_key_len);
+      std::string i_str = absl::StrCat("", i);
```
You can also use `absl::SimpleAtoi` here. Also, while you are here please constify all variables that you can.

inf-rno commented:
Addressed the comments and updated the PR. I added a `ConsistentHashingLbConfig` within the `CommonLbConfig`. This gave us the best of most things: proper namespacing, hierarchy, no duplication all without changing the code too much. Please take another look.

mattklein123 commented on docs/root/intro/version_history.rst:
```diff
@@ -31,6 +31,7 @@ Version history
 * http: added :ref:`max_stream_duration <envoy_api_field_core.HttpProtocolOptions.max_stream_duration>` to specify the duration of existing streams. See :ref:`connection and stream timeouts <faq_configuration_timeouts>`.
 * listener filters: listener filter extensions use the "envoy.filters.listener" name space. A
   mapping of extension names is available in the :ref:`deprecated <deprecated>` documentation.
+* loadbalancing: added support for using hostname for consistent hash loadbalancing via :ref:`consistent_hash_lb_config <envoy_api_field_cluster.CommonLbConfig.ConsistentHashingLbConfig>`.
```
nit: alpha order

inf-rno commented:
You're right. Even though it would work with LogicalDNS and just one IP address per hostname it is a misuse of this option so I've disabled it and fixed the nit. Please have another look.

mattklein123 commented:
Thanks LGTM. Can you check docs CI?

/wait

mattrobenolt commented:
Hey @inf-rno, wanted to follow up on some questions about how to correctly use this or the use cases. I'm struggling to pair it with how I'd expect to use it, and this could be me being naive and trying to introspect this behavior well.

I would expect to be able to use this feature with a CNAME record coupled with service discovery, since this is in reference to the original issue that mentions kubernetes, consul, etc. They usually surface the "service" behind a CNAME, that then fans out into N CNAME and following A records. This seems counter to the documentation that explicitly says this only applies to a DNS record that returns 1 IP.

So this seems to imply I'd need to use a config such as:

```yaml
  - name: upstream
    type: strict_dns
    lb_policy: MAGLEV
    common_lb_config:
      consistent_hashing_lb_config:
        use_hostname_for_hashing: true
    load_assignment:
      cluster_name: upstream
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: service-host-1
                port_value: 1234
        - endpoint:
            address:
              socket_address:
                address: service-host-2
                port_value: 1234
```

Which requires statically listing all of the known hosts. vs doing something like:

```yaml
  - name: upstream
    type: strict_dns
    lb_policy: MAGLEV
    common_lb_config:
      consistent_hashing_lb_config:
        use_hostname_for_hashing: true
    load_assignment:
      cluster_name: upstream
      endpoints:
      - lb_endpoints:
        - endpoint:
            address:
              socket_address:
                address: service  # service resolves to `service-host-*`
                port_value: 1234
```

Is this correct? Struggling to figure out how to reliably get consistent hashing while using service discovery. My particular use case is a ReplicaSet in Kubernetes without having my own manager that writes out config and updates envoy on changes.

inf-rno commented:
@mattrobenolt You're right this is designed to only work with the first config example you've mentioned where you list out all the members of your StatefulSet as explicit endpoints. I'm not sure I understand your second use case within a hostname based consistent hashing use case. Are you expecting envoy to traverse the CNAME indirections (over multiple levels?) to automatically detect and configure the upstreams or something like that?

mattrobenolt commented:
Yes, that’s correct. Both Consul and Kubernetes work by giving you a CNAME which points to a list of other FQDN CNAME records. Similar to an SRV record. So given this, you could first query the CNAME, then use the hostnames returned as the endpoint records, then further resolving those into A records. I think this makes it a tad over complicated, but I’m pretty sure with the response of the first CNAME all the data is returned. I don’t know enough offhand about DNS protocols, but I’ll investigate tomorrow.

mattrobenolt commented:
I guess alternatively, just supporting SRV records would simplify this and be more explicit in behavior.
